File: webapp/packages/desktop/src/stores/menuVisibility.ts
import { defineStore } from 'pinia'
import { ApiError } from '@ct/shared'
import { client } from '../api'

export const menuGroups = [
  { title: '总览', items: [['/', '总览']] },
  { title: '监控分析', items: [['/customers', '客户监控'], ['/channels', '渠道监控'], ['/models', '模型监控'], ['/runtime', '系统状态']] },
  { title: '数据查询', items: [['/usage', '用量统计'], ['/readonly-users', '用户管理'], ['/readonly-logs', '使用日志'], ['/container-logs', '容器日志']] },
  { title: '账单管理', items: [['/billing', '用户账单'], ['/billing/channels', '上游账单'], ['/billing/tasks', '账单任务'], ['/billing/discounts', '渠道折扣']] },
  { title: '系统管理', items: [['/tuning', '调权中心'], ['/alerts', '告警中心'], ['/notifications', '通知设置'], ['/instances', '实例管理'], ['/log-archives', '日志归档'], ['/access-users', '账号管理'], ['/models/manage', '模型广场'], ['/billing/upstreams', '上游管理'], ['/settings', '系统设置'], ['/audits', '操作审计']] },
] as const
type Response = { items: Record<string, boolean> }
let pending: Promise<void> | null = null
export const useMenuVisibilityStore = defineStore('menuVisibility', {
  state: () => ({ items: {} as Record<string, boolean>, ready: false, supported: true, loading: false, error: '', revision: 0 }),
  getters: { visible: state => (path: string) => state.ready && state.items[path] !== false },
  actions: {
    async load() {
      if (pending) return pending
      this.loading = true
      const revision = this.revision
      pending = (async () => {
      try {
        const result = await client.request<Response>('/api/dashboard/menu-visibility')
        if (revision !== this.revision) return
        this.items = result.items; this.ready = true; this.supported = true; this.error = ''
      } catch (error) {
        if (revision !== this.revision) return
        if (error instanceof ApiError && error.status === 404) {
          this.items = {}; this.ready = true; this.supported = false
          this.error = '当前后端尚未支持菜单显示设置，请升级 Server 后使用。'
        } else { this.error = '菜单配置加载失败，请重试。' }
      } finally { this.loading = false }
      })()
      try { await pending } finally { pending = null }
    },
    async save(items: Record<string, boolean>) {
      const result = await client.request<Response>('/api/dashboard/menu-visibility', { method: 'PUT', body: JSON.stringify({ items }) })
      this.revision++
      this.items = result.items; this.ready = true; this.error = ''
    },
  },
})

File: server/internal/dashboard/menu_visibility.go
package dashboard

import (
	ctauth "controltower/server/internal/auth"
	"controltower/server/internal/storage"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"io"
	"net/http"
	"time"
)

type MenuVisibilityStore interface {
	MenuVisibility() (string, error)
	SaveMenuVisibility(string, string, time.Time) error
	InsertOperationAudit(storage.OperationAudit) error
}

// These are sidebar entry IDs, not access permissions. Missing entries are visible.
var menuPaths = map[string]bool{
	"/": true, "/customers": true, "/channels": true, "/models": true, "/runtime": true,
	"/usage": true, "/readonly-users": true, "/readonly-logs": true, "/container-logs": true,
	"/billing": true, "/billing/channels": true, "/billing/tasks": true, "/billing/discounts": true,
	"/tuning": true, "/alerts": true, "/notifications": true, "/instances": true, "/log-archives": true,
	"/access-users": true, "/models/manage": true, "/billing/upstreams": true, "/settings": true, "/audits": true,
}

type MenuVisibilityHandler struct{ Store MenuVisibilityStore }

func (h MenuVisibilityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Cache-Control", "no-store")
	if r.Method != http.MethodGet && r.Method != http.MethodPut {
		writeDashboardError(w, 405, "method_not_allowed")
		return
	}
	if r.Method == http.MethodPut {
		// The dashboard bearer token is trusted; sessions must own settings.manage.
		if user, ok := ctauth.CurrentUser(r); ok && !ctauth.HasPermission(user, "settings.manage") {
			writeDashboardError(w, 403, "forbidden")
			return
		}
	}
	raw, err := h.Store.MenuVisibility()
	if err != nil {
		writeDashboardError(w, 500, "query_failed")
		return
	}
	values := map[string]bool{}
	if err := json.Unmarshal([]byte(raw), &values); err != nil || values == nil {
		writeDashboardError(w, 500, "invalid_stored_menu_visibility")
		return
	}
	if r.Method == http.MethodPut {
		var req struct {
			Items map[string]*bool `json:"items"`
		}
		decoder := json.NewDecoder(http.MaxBytesReader(w, r.Body, 8192))
		decoder.DisallowUnknownFields()
		if err := decoder.Decode(&req); err != nil || req.Items == nil {
			writeDashboardError(w, 400, "invalid_menu_visibility")
			return
		}
		var extra any
		if decoder.Decode(&extra) != io.EOF {
			writeDashboardError(w, 400, "invalid_json")
			return
		}
		next := map[string]bool{}
		for path, visible := range req.Items {
			if !menuPaths[path] || visible == nil {
				writeDashboardError(w, 400, "invalid_menu_visibility")
				return
			}
			next[path] = *visible
		}
		encoded, _ := json.Marshal(next)
		actor := ctauth.Actor(r)
		if err := h.Store.SaveMenuVisibility(string(encoded), actor, time.Now().UTC()); err != nil {
			writeDashboardError(w, 500, "update_failed")
			return
		}
		id := make([]byte, 16)
		_, _ = rand.Read(id)
		if err := h.Store.InsertOperationAudit(storage.OperationAudit{ID: hex.EncodeToString(id), OperationType: "menu_visibility.update", TargetType: "system_settings", TargetID: "global", ActorID: actor, BeforeSummary: raw, AfterSummary: string(encoded), Status: "succeeded", CreatedAt: time.Now().UTC()}); err != nil {
			writeDashboardError(w, 500, "audit_failed")
			return
		}
		values = next
	}
	writeDashboardJSON(w, 200, map[string]any{"items": values})
}
